src/detect-bytemath.c

```diff
@@ -0,0 +1,1196 @@
+ */
+void DetectBytemathRegister(void)
+{
+    sigmatch_table[DETECT_BYTEMATH].name = "byte_math";
+    sigmatch_table[DETECT_BYTEMATH].Match = NULL;
+    sigmatch_table[DETECT_BYTEMATH].Setup = DetectByteMathSetup;
+    sigmatch_table[DETECT_BYTEMATH].Free = DetectByteMathFree;
+    sigmatch_table[DETECT_BYTEMATH].RegisterTests = DetectByteMathRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+}
+
+int DetectByteMathDoMatch(DetectEngineThreadCtx *det_ctx, const SigMatchData *smd,
+                          const Signature *s, const uint8_t *payload,
+                          uint16_t payload_len, uint64_t rvalue, uint64_t *value, uint8_t endian)
+{
+    DetectByteMathData *data = (DetectByteMathData *)smd->ctx;
+    const uint8_t *ptr;
+    int32_t len;
+    uint64_t val;
+    int extbytes;
+
+    if (payload_len == 0) {
+        return 0;
+    }
+
+    /* Calculate the ptr value for the byte-math op and length remaining in
+     * the packet from that point.
+     */
+    if (data->flags & DETECT_BYTEMATH_FLAG_RELATIVE) {
+        SCLogDebug("relative, working with det_ctx->buffer_offset %"PRIu32", "
+                   "data->offset %"PRIu32"", det_ctx->buffer_offset, data->offset);
+
+        ptr = payload + det_ctx->buffer_offset;
+        len = payload_len - det_ctx->buffer_offset;
+
+        ptr += data->offset;
+        len -= data->offset;
+
+        /* No match if there is no relative base */
+        if (len <= 0) {
+            return 0;
+        }
+    } else {
+        SCLogDebug("absolute, data->offset %"PRIu32"", data->offset);
+
+        ptr = payload + data->offset;
+        len = payload_len - data->offset;
+    }
+
+    /* Validate that the to-be-extracted is within the packet */
+    if (ptr < payload || data->nbytes > len) {
+        SCLogDebug("Data not within payload pkt=%p, ptr=%p, len=%"PRIu32", nbytes=%d",
+                    payload, ptr, len, data->nbytes);
+        return 0;
+    }
+
+    /* Extract the byte data */
+    if (data->flags & DETECT_BYTEMATH_FLAG_STRING) {
+        extbytes = ByteExtractStringUint64(&val, data->base,
+                                           data->nbytes, (const char *)ptr);
+        if (extbytes <= 0) {
+            if (val == 0) {
+                SCLogDebug("No Numeric value");
+                return 0;
+            } else {
+                SCLogDebug("error extracting %d bytes of string data: %d",
+                           data->nbytes, extbytes);
+                return -1;
+            }
+        }
+    } else {
+        int endianness = (endian == DETECT_BYTEMATH_ENDIAN_BIG) ?
+                          BYTE_BIG_ENDIAN : BYTE_LITTLE_ENDIAN;
+        extbytes = ByteExtractUint64(&val, endianness, data->nbytes, ptr);
+        if (extbytes != data->nbytes) {
+            SCLogDebug("error extracting %d bytes of numeric data: %d",
+                       data->nbytes, extbytes);
+            return 0;
+        }
+    }
+
+    if (extbytes > len) {
```

Comment:
is this a condition that can be reached? It seems more like a bug to me. If it is, we should probably use BUG_ON?